apps/prairielearn/src/ee/pages/studentCourseInstanceUpgrade/studentCourseInstanceUpgrade.ts

```diff
@@ -17,7 +50,238 @@ router.get(
+          quantity: 1,
+        });
+      }
+
+      // Validate that the plan names from the client are actually valid. We
+      // consider them to be valid if they are in the list of missing plans,
+      // which in turn is defined as a plan that is required for the current
+      // course instance and isn't already granted to the user.
+      const planGrants = await getPlanGrantsForPartialContexts({
+        institution_id: institution.id,
+        course_instance_id: course_instance.id,
+        user_id: user.user_id,
+      });
+      const requiredPlans = await getRequiredPlansForCourseInstance(res.locals.course_instance.id);
+      const missingPlans = getMissingPlanGrants(planGrants, requiredPlans);
+      if (!planNames.every((planName) => missingPlans.includes(planName))) {
+        throw error.make(400, 'Invalid plan selection.');
+      }
+
+      const host = getCanonicalHost(req);
+      const urlBase = `${host}/pl/course_instance/${course_instance.id}/upgrade`;
+
+      const stripe = getStripeClient();
+      const customerId = await getOrCreateStripeCustomerId(user.user_id, {
+        name: user.name,
+      });
+      const session = await stripe.checkout.sessions.create({
+        customer: customerId,
+        customer_update: {
+          name: 'auto',
+          address: 'auto',
+        },
+        metadata: {
+          prairielearn_institution_id: institution.id,
+          prairielearn_institution_name: `${institution.long_name} (${institution.short_name})`,
+          prairielearn_course_id: course.id,
+          prairielearn_course_name: `${course.short_name}: ${course.title}`,
+          prairielearn_course_instance_id: course_instance.id,
+          prairielearn_course_instance_name: `${course_instance.long_name} (${course_instance.short_name})`,
+          prairielearn_user_id: user.user_id,
+        },
+        line_items: lineItems,
+        mode: 'payment',
+        success_url: `${urlBase}/success?session_id={CHECKOUT_SESSION_ID}`,
+        cancel_url: urlBase,
+      });
+
+      await insertStripeCheckoutSessionForUserInCourseInstance({
+        stripe_object_id: session.id,
+        institution_id: institution.id,
+        course_instance_id: course_instance.id,
+        user_id: user.user_id,
+        data: session,
+        plan_names: planNames,
+      });
+
+      if (!session.url) throw error.make(500, 'Stripe session URL not found');
+
+      res.redirect(session.url);
+    } else {
+      throw error.make(400, `Unknown action: ${req.body.__action}`);
+    }
+  }),
+);
+
+router.get(
+  '/success',
+  asyncHandler(async (req, res) => {
+    const institution = InstitutionSchema.parse(res.locals.institution);
+    const course = CourseSchema.parse(res.locals.course);
+    const course_instance = CourseInstanceSchema.parse(res.locals.course_instance);
+    const authn_user = UserSchema.parse(res.locals.authn_user);
+
+    if (!req.query.session_id) throw error.make(400, 'Missing session_id');
```

Comment:
Should `session_id` here now be called `stripe_object_id`? And similarly below.

Reply:
I think it doesn't really matter; this is the query param that Stripe passes back to us, I copied this straight out of their docs.